iree/hal/cts/command_buffer_test.cc

```diff
@@ -0,0 +1,284 @@
+
+  EXPECT_TRUE((iree_hal_command_buffer_allowed_categories(command_buffer) &
+               IREE_HAL_COMMAND_CATEGORY_DISPATCH) ==
+              IREE_HAL_COMMAND_CATEGORY_DISPATCH);
+
+  iree_hal_command_buffer_release(command_buffer);
+}
+
+TEST_P(CommandBufferTest, BeginEnd) {
+  iree_hal_command_buffer_t* command_buffer;
+  IREE_ASSERT_OK(iree_hal_command_buffer_create(
+      device_, IREE_HAL_COMMAND_BUFFER_MODE_ONE_SHOT,
+      IREE_HAL_COMMAND_CATEGORY_DISPATCH, &command_buffer));
+
+  IREE_EXPECT_OK(iree_hal_command_buffer_begin(command_buffer));
+  IREE_EXPECT_OK(iree_hal_command_buffer_end(command_buffer));
+
+  iree_hal_command_buffer_release(command_buffer);
+}
+
+TEST_P(CommandBufferTest, FillBufferWithRepeatedBytes) {
+  iree_hal_command_buffer_t* command_buffer;
+  IREE_ASSERT_OK(iree_hal_command_buffer_create(
+      device_, IREE_HAL_COMMAND_BUFFER_MODE_ONE_SHOT,
+      IREE_HAL_COMMAND_CATEGORY_TRANSFER, &command_buffer));
+
+  iree_hal_buffer_t* device_buffer;
+  IREE_ASSERT_OK(iree_hal_allocator_allocate_buffer(
+      device_allocator_,
+      IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL | IREE_HAL_MEMORY_TYPE_HOST_VISIBLE,
+      IREE_HAL_BUFFER_USAGE_ALL, kBufferSize, &device_buffer));
+
+  std::vector<uint8_t> reference_buffer(kBufferSize);
+
+  IREE_ASSERT_OK(iree_hal_command_buffer_begin(command_buffer));
+
+  // Fill the device buffer with segments of different values so that we can
+  // test both fill and offset/size.
+  uint8_t val1 = 0x07;
+  iree_hal_command_buffer_fill_buffer(
+      command_buffer, device_buffer,
+      /*target_offset=*/0, /*length=*/kBufferSize / 4, /*pattern=*/&val1,
+      /*pattern_length=*/1);
+  std::memset(reference_buffer.data(), val1, kBufferSize / 4);
+
+  uint8_t val2 = 0xbe;
+  iree_hal_command_buffer_fill_buffer(command_buffer, device_buffer,
+                                      /*target_offset=*/kBufferSize / 4,
+                                      /*length=*/kBufferSize / 4,
+                                      /*pattern=*/&val2,
+                                      /*pattern_length=*/1);
+  std::memset(reference_buffer.data() + kBufferSize / 4, val2, kBufferSize / 4);
+
+  uint8_t val3 = 0x54;
+  iree_hal_command_buffer_fill_buffer(command_buffer, device_buffer,
+                                      /*target_offset=*/kBufferSize / 2,
+                                      /*length=*/kBufferSize / 2,
+                                      /*pattern=*/&val3,
+                                      /*pattern_length=*/1);
+  std::memset(reference_buffer.data() + kBufferSize / 2, val3, kBufferSize / 2);
+
+  IREE_ASSERT_OK(iree_hal_command_buffer_end(command_buffer));
+
+  iree_hal_semaphore_t* signal_semaphore;
+  IREE_ASSERT_OK(iree_hal_semaphore_create(device_, 0u, &signal_semaphore));
+  SubmitAndWait(command_buffer, IREE_HAL_COMMAND_CATEGORY_TRANSFER,
+                signal_semaphore);
+
+  // Map the device buffer and compare.
+  iree_hal_buffer_mapping_t mapped_memory;
```

Comment:
thinking about what we were talking about with fill/read/write/copy buffer functions: you'll want to keep this as iree_hal_buffer_read_data today - we can then make that handle both cases of mappable vs. unmappable memory (if mappable do what it does today, like this, and if unmappable then create the command buffer and issue) - since mappable memory is not a requirement we want to avoid mapping in tests unless the tests are explicitly testing mapping

Reply:
SG. Mapping was used in the prior code (sort of... lol). I've updated to use `iree_hal_buffer_read_data`. Separately, I'd also like to add coverage for mapping (where available).

I have some thoughts on how to exercise optional components in the CTS, like checking `iree_hal_allocator_query_buffer_compatibility` for `IREE_STATUS_UNAVAILABLE` and skipping test cases that use mapping.